### PR TITLE
Add p2p crawler cmd

### PR DESCRIPTION
**Pull request by bCoder778**

### CODE REVIEW

File: cmd/crawler/node/node.go [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+	}
+	gs.Tips = append(gs.Tips, &pb.Hash{Hash: genesisHash.Bytes()})
+
+	resp := &pb.ChainState{
+		GenesisHash:     &pb.Hash{Hash: genesisHash.Bytes()},
+		ProtocolVersion: pv.ProtocolVersion,
+		Timestamp:       uint64(roughtime.Now().Unix()),
+		Services:        uint64(pv.Relay),
+		GraphState:      gs,
+		UserAgent:       []byte("qitmeer-relay"),
```
Comment (dindinw) [inline cmd/crawler/node/node.go:292]: should not use the same agent name with `qitmeer-relay`

File: cmd/crawler/node/node.go [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+	return nil
+}
+
+func (node *Node) connectWithNewPeers() {
+	ti := time.NewTicker(30 * time.Second)
+	defer ti.Stop()
+
+	for {
+		select {
+		case <-ti.C:
```
Comment (dindinw) [inline cmd/crawler/node/node.go:173]: I think you need to stop/shutdown your goroutine gracefully in your `select`

File: cmd/crawler/node/node.go [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+				if err != nil {
+					log.Log.Trace(fmt.Sprintf("Wrong remote address %s to p2p address, %v", getConnPeerAddress(p.Conn), err))
+					continue
+				}
+				addrInfo, err := peer.AddrInfoFromP2pAddr(maAddr)
+				if err != nil {
+					log.Log.Trace(fmt.Sprintf("Wrong remote address %s to p2p address, %v", p.Conn.RemoteMultiaddr().String(), err))
+					continue
+				}
+				go func(info peer.AddrInfo) {
```
Comment (dindinw) [inline cmd/crawler/node/node.go:186]: stop your goroutine gracefully.

File: cmd/crawler/node/node.go [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+
+func (node *Node) connectWithAllPeers(multiAddrs []multiaddr.Multiaddr) {
+	addrInfos, err := peer.AddrInfosFromP2pAddrs(multiAddrs...)
+	if err != nil {
+		log.Log.Error(fmt.Sprintf("Could not convert to peer address info's from multiaddresses: %v", err))
+		return
+	}
+	for _, info := range addrInfos {
+		// make each dial non-blocking
+		go func(info peer.AddrInfo) {
```
Comment (dindinw) [inline cmd/crawler/node/node.go:210]: stop your goroutine gracefully.

**Review by dindinw — CHANGES_REQUESTED** [review]
Please double-check how to shutdown your go-routine gracefully.

File: cmd/crawler/node/node.go (hunk 3 above)
Comment (bCoder778) [reply cmd/crawler/node/node.go:186]: ok

File: cmd/crawler/node/config.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/Qitmeer/qitmeer/common/util"
+	"github.com/Qitmeer/qitmeer/params"
+	"github.com/urfave/cli/v2"
+	"os"
+	"path/filepath"
+)
+
+const (
+	defaultDataDirname = "crawler-data"
+	defaultPort        = "1001"
```
Comment (lochjin) [inline cmd/crawler/node/config.go:14]: Please use a port greater than 1024.

File: cmd/crawler/node/node.go [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+	}
+
+	srcMAddr, err := multiaddr.NewMultiaddr(fmt.Sprintf("/ip4/%s/tcp/%s", defaultIP, node.cfg.Port))
+	if err != nil {
+		log.Log.Error("Unable to construct multiaddr %v", err)
+		return err
+	}
+
+	opts := []libp2p.Option{
+		libp2p.EnableRelay(relay.OptHop),
```
Comment (lochjin) [inline cmd/crawler/node/node.go:107]: If it's just crawler, `EnableRelay` should be removed

File: cmd/crawler/node/node.go [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+		MainOrder:  0,
+		Tips:       []*pb.Hash{},
+	}
+	gs.Tips = append(gs.Tips, &pb.Hash{Hash: genesisHash.Bytes()})
+
+	resp := &pb.ChainState{
+		GenesisHash:     &pb.Hash{Hash: genesisHash.Bytes()},
+		ProtocolVersion: pv.ProtocolVersion,
+		Timestamp:       uint64(roughtime.Now().Unix()),
+		Services:        uint64(pv.Relay),
```
Comment (lochjin) [inline cmd/crawler/node/node.go:306]: You can use `pv.Observer`

**Review by lochjin — CHANGES_REQUESTED** [review]

File: cmd/crawler/node/node.go [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+	return typeAssertedKey
+}
+
+func convertToInterfacePubkey(pubkey *ecdsa.PublicKey) crypto.PubKey {
+	typeAssertedKey := crypto.PubKey((*crypto.Secp256k1PublicKey)((*secp256k1.PublicKey)(pubkey)))
+	return typeAssertedKey
+}
+
+func getConnPeerAddress(conn network.Conn) string {
+	return fmt.Sprintf("%s/ipfs/%s", conn.RemoteMultiaddr().String(), conn.RemotePeer())
```
Comment (dindinw) [inline cmd/crawler/node/node.go:432]: why ipfs?

Comment (bCoder778) [reply cmd/crawler/node/node.go:432]: If I understand correctly, Multiaddr needs to be formed in this form, and ipfs is a fixed field

File: cmd/crawler/node/config.go (hunk 5 above)
Comment (bCoder778) [reply cmd/crawler/node/config.go]: ok

File: cmd/crawler/node/node.go (hunk 8 above)
Comment (dindinw) [reply cmd/crawler/node/node.go:431]: https://github.com/multiformats/multiaddr has no requirement to fix to `ipfs` prefix, and we use `p2p` internally at https://github.com/Qitmeer/qitmeer/blob/0.10-dev/p2p/addrfactory.go#L134

Comment (bCoder778) [reply cmd/crawler/node/node.go]: ok

**Review by dindinw — APPROVED** [review]

**Review by lochjin — APPROVED** [review]

**Pull request merged**